pkg/mcp/internal/queue.go

```diff
@@ -0,0 +1,175 @@
+//   state of the queue and should stop checking Ready() and invoking Dequeue() when
+//   the channel returned by Done() is closed.
+//
+// This is intended to be used by the MCP source/server packages for managing
+// per-type watch state.
+type UniqueQueue struct {
+	sync.Mutex
+	doneChan       chan struct{}
+	doneChanClosed bool
+	queuedSet      map[interface{}]struct{}
+	// Enqueue at the tail, Dequeue from the head
+	// head == tail   => empty
+	// head == tail+1 => full
+	head     int
+	tail     int
+	queue    []interface{}
+	maxDepth int
+
+	scheduleChan chan struct{}
+}
+
+// NewUniqueScheduledQueue creates a new unique queue specialized for MCP source/server implementations.
+func NewUniqueScheduledQueue(maxDepth int) *UniqueQueue {
+	return &UniqueQueue{
+		queue:        make([]interface{}, maxDepth+1),
+		queuedSet:    make(map[interface{}]struct{}, maxDepth),
+		maxDepth:     maxDepth,
+		scheduleChan: make(chan struct{}, maxDepth),
+		doneChan:     make(chan struct{}),
+	}
+}
+
+// return the next index accounting for wrap
+func (q *UniqueQueue) inc(idx int) int {
+	idx = idx + 1
+	if idx == q.maxDepth+1 {
+		idx = 0
+	}
+	return idx
+}
+
+// Empty returns true if the queue is empty
+func (q *UniqueQueue) Empty() bool {
+	q.Lock()
+	defer q.Unlock()
+	return q.head == q.tail
+}
+
+// internal version of full() that can be called
+// with the lock held.
+func (q *UniqueQueue) full() bool {
+	return q.head == q.inc(q.tail)
+}
+
+// Full returns true if the queue is full
+func (q *UniqueQueue) Full() bool {
+	q.Lock()
+	defer q.Unlock()
+	return q.full()
+}
+
+// Enqueue an item in the queue. The same item may be safely enqueued multiple
+// times. Attempts to enqueue an already queued item have no affect on the order
+// of already queued items.
+//
+// Returns true if the item exists in the queue upon return. Otherwise,
+// returns false if the item could not be queued.
+func (q *UniqueQueue) Enqueue(w interface{}) bool {
+	q.Lock()
+	defer q.Unlock()
+
+	// already in the queued set of items
+	if _, ok := q.queuedSet[w]; ok {
+		return true
+	}
+
+	if q.full() {
+		return false
+	}
+
+	q.queuedSet[w] = struct{}{}
+	q.queue[q.tail] = w
+	q.tail = q.inc(q.tail)
+
+	q.Schedule()
+	return true
+}
+
+func (q *UniqueQueue) Dequeue() interface{} {
+	q.Lock()
+	defer q.Unlock()
+
+	if q.head == q.tail {
+		return nil
+	}
+
+	w := q.queue[q.head]
+	q.head = q.inc(q.head)
+	delete(q.queuedSet, w)
+	return w
+}
+
+func (q *UniqueQueue) Ready() <-chan struct{} {
+	return q.scheduleChan
+}
+
+func (q *UniqueQueue) Done() <-chan struct{} {
+	return q.doneChan
+}
+
+// Controls the delay for re-retrying a configuration push if the previous
+// attempt was not possible, e.g. the lower-level serving layer was busy. This
+// should typically be set fairly small (order of milliseconds).
+var retryPushDelay = env.Duration("RETRY_PUSH_DELAY", 10*time.Millisecond)
+
+func (q *UniqueQueue) Schedule() {
+	select {
+	case q.scheduleChan <- struct{}{}:
```

Comment:
Doesn't this imply that the max reached? It normally shouldn't happen, but you wouldn't need to do a retry push?

Reply:
This would occur if the caller drained the queue (via `Dequeue()`) without first reading from the ready channel (via `Ready()`). I'll add a comment to `Dequeue` to make this requirement clearer.